lib/AST/TypeCheckRequests.cpp

```diff
@@ -2075,3 +2075,62 @@ void SemanticDeclAttrsRequest::cacheResult(DeclAttributes attrs) const {
   auto decl = std::get<0>(getStorage());
   const_cast<Decl *>(decl)->setSemanticAttrsComputed(true);
 }
+
+//----------------------------------------------------------------------------//
+// UniqueUnderlyingTypeSubstitutionsRequest computation.
+//----------------------------------------------------------------------------//
+
+static bool
+shouldTypecheckFunctionBodyForUniqueUnderlyingType(AbstractFunctionDecl *afd) {
+  auto mod = afd->getModuleContext();
+  if (!mod->isMainModule())
+    return true;
+
+  // If the main module has no primary source files then the compilation is a
+  // whole module build and all source files can be typechecked.
+  if (mod->getPrimarySourceFiles().size() == 0)
+    return true;
+
+  auto sf = afd->getParentSourceFile();
+  if (!sf)
+    return true;
+
+  if (sf->getKind() != FileUnitKind::Source)
+    return true;
+
+  if (sf->isPrimary())
+    return true;
+
+  return false;
+}
+
+llvm::Optional<SubstitutionMap>
+UniqueUnderlyingTypeSubstitutionsRequest::evaluate(
+    Evaluator &evaluator, const OpaqueTypeDecl *decl) const {
+  auto namingDecl = decl->getNamingDecl();
+
+  // If the naming declaration is a function, typecheck its body. This will
+  // have a side-effect of setting UniqueUnderlyingType on the opaque type.
```

Comment:
What if the naming decl is a property or subscript? I believe the same logic should apply

Reply:
Yup good catch